src/realm/sync/network/default_socket.cpp

```diff
@@ -65,25 +65,90 @@ class DefaultWebSocketImpl final : public WebSocketInterface, public Config {
     void websocket_read_error_handler(std::error_code ec) override
     {
         m_logger.error("Reading failed: %1", ec.message()); // Throws
-        m_observer.websocket_read_or_write_error_handler(ec);
+        constexpr bool was_clean = false;
+        websocket_error_and_close_handler(was_clean, Status{ErrorCodes::ReadError, ec.message()});
     }
     void websocket_write_error_handler(std::error_code ec) override
     {
         m_logger.error("Writing failed: %1", ec.message()); // Throws
-        m_observer.websocket_read_or_write_error_handler(ec);
+        constexpr bool was_clean = false;
+        websocket_error_and_close_handler(was_clean, Status{ErrorCodes::WriteError, ec.message()});
     }
     void websocket_handshake_error_handler(std::error_code ec, const HTTPHeaders*,
                                            const std::string_view* body) override
     {
-        m_observer.websocket_handshake_error_handler(ec, body);
+        ErrorCodes::Error error;
+        bool was_clean;
+
+        if (ec == websocket::Error::bad_response_3xx_redirection ||
```

Comment:
Use the abnormal closure error code and was_clean = false for generic 3XX HTTP responses, since we only want to specifically handle 301 or 308 as a redirect (technically we'll handle it the same way, but that will change in the future).

Reply:
I'll use the new `Retry` with was_clean = false. Since we want to differentiate it from 301, we should not refresh the token then (we do that for abnormal closure).